# the exon.legnth.file is generated by /bar/cfan/python_scripts/TPM/exon_length.py
tpm <- function(count.table, exon.length) {
  # count.table and exon.length can be strings for file name or dataframe/tibble
  if (is.character(count.table))
    count.table <- read.table(count.table, as.is = T, header = T)
  if (!is.matrix(count.table)) {
    rownames(count.table) <- count.table[,1]
    count.table <- count.table[,-1, drop = FALSE] %>% as.matrix()
  }

  if (is.character(exon.length))
    exon.length <- read.table(exon.length, as.is = T, header = F)
  colnames(exon.length) <- c("gene", "length")
  #return(exon.length)
  genes <- rownames(count.table)
  # print(genes)
  tpm <- apply(count.table, 2, function (x) {
    df <- data.frame(gene = genes, counts = x) %>%
      left_join(exon.length)
    # print(head(df))
    RPK <- df$counts/(df$length/1000)
    PerM <- sum(RPK)/1000000
    tpm.col <- RPK/PerM
    return(tpm.col)
  })
  rownames(tpm) <- genes
  return(tpm)
}

# t <- tpm("~/4dn/nk/oshea/rna_fanc/oshea_rna.renamed", exon.length = "~/genomes/mm10/gencode/gencode.vM24.exon.length.tsv")
# t %>% str()
#
# A <- data.frame(gene=c("a", "b", "c"), sample1 = c(1000,2000,3000))
# E <- data.frame(gene = c("a", "b", "c"), length = c(10 ,20, 30))
# toy <- tpm(A, E)
